esutil/bulk_indexer.go

```diff
@@ -0,0 +1,509 @@
+		if w.buf.Len() > 0 {
+			if err := w.flush(); err != nil {
+				// TODO(karmi): Wrap error
+				w.mu.Unlock()
+				return fmt.Errorf("%s", err)
+			}
+			w.mu.Unlock()
+		}
+	}
+
+	return nil
+}
+
+// Stats returns indexer statistics.
+//
+func (bi *bulkIndexer) Stats() BulkIndexerStats {
+	return BulkIndexerStats{
+		NumAdded:    uint(atomic.LoadUint64(&bi.stats.numAdded)),
+		NumFlushed:  uint(atomic.LoadUint64(&bi.stats.numFlushed)),
+		NumFailed:   uint(atomic.LoadUint64(&bi.stats.numFailed)),
+		NumIndexed:  uint(atomic.LoadUint64(&bi.stats.numIndexed)),
+		NumCreated:  uint(atomic.LoadUint64(&bi.stats.numCreated)),
+		NumUpdated:  uint(atomic.LoadUint64(&bi.stats.numUpdated)),
+		NumDeleted:  uint(atomic.LoadUint64(&bi.stats.numDeleted)),
+		NumRequests: uint(atomic.LoadUint64(&bi.stats.numRequests)),
+	}
+}
+
+// init initializes the bulk indexer.
+//
+func (bi *bulkIndexer) init() {
+	bi.queue = make(chan BulkIndexerItem, bi.config.NumWorkers)
+
+	for i := 1; i <= bi.config.NumWorkers; i++ {
+		w := worker{
+			id:  i,
+			ch:  bi.queue,
+			bi:  bi,
+			buf: bytes.NewBuffer(make([]byte, 0, bi.config.FlushBytes)),
+			aux: make([]byte, 0, 512)}
+		w.run()
+		bi.workers = append(bi.workers, &w)
+	}
+	bi.wg.Add(bi.config.NumWorkers)
+
+	bi.ticker = time.NewTicker(bi.config.FlushInterval)
+	go func() error {
+		for {
+			select {
+			case <-bi.ticker.C:
+				if bi.config.DebugLogger != nil {
+					bi.config.DebugLogger.Printf("[indexer] Auto-flushing workers after %s\n", bi.config.FlushInterval)
+				}
+				for _, w := range bi.workers {
+					w.mu.Lock()
+					if w.buf.Len() > 0 {
+						if err := w.flush(); err != nil {
+							// TODO(karmi): Wrap error
+							w.mu.Unlock()
+							return fmt.Errorf("%s", err)
+						}
+					}
+					w.mu.Unlock()
+				}
+			}
+		}
+	}()
+}
+
+// worker represents an indexer worker.
+//
+type worker struct {
+	id    int
+	ch    <-chan BulkIndexerItem
+	mu    sync.Mutex
+	bi    *bulkIndexer
+	buf   *bytes.Buffer
+	aux   []byte
+	items []BulkIndexerItem
+
+	isFlushing bool
+}
+
+// run launches the worker in a goroutine.
+//
+func (w *worker) run() {
+	go func() {
+		if w.bi.config.DebugLogger != nil {
+			w.bi.config.DebugLogger.Printf("[worker-%03d] Started\n", w.id)
+		}
+		defer w.bi.wg.Done()
+
+		for item := range w.ch {
+			w.mu.Lock()
+
+			if w.bi.config.DebugLogger != nil {
+				w.bi.config.DebugLogger.Printf("[worker-%03d] Received item [%s:%s]\n", w.id, item.Action, item.DocumentID)
+			}
+
+			if err := w.writeMeta(item); err != nil {
+				if item.OnFailure != nil {
+					item.OnFailure(item, BulkIndexerResponseItem{}, err)
+				}
+				atomic.AddUint64(&w.bi.stats.numFailed, 1)
+				w.mu.Unlock()
+				continue
+			}
+
+			if err := w.writeBody(item); err != nil {
+				if item.OnFailure != nil {
+					item.OnFailure(item, BulkIndexerResponseItem{}, err)
+				}
+				atomic.AddUint64(&w.bi.stats.numFailed, 1)
+				w.mu.Unlock()
+				continue
+			}
+
+			w.items = append(w.items, item)
```

Comment:
It would to avoid appending to `w.items` if there's no callback(s) defined. That would require you to record the index on the items, for matching up with the results.

Reply:
Sure, this doesn't need to be done now. I'll stick to reviewing things that affect the external API then.

Reply:
Please keep raising the questions, though! Even though all the suggestions might not be implemented right away, it's good to keep track of them.

Reply:
The items will be stored on heap. Due to the flush you reuse the items array, making it a rather cheap operation (no extra allocation). Drawback is that you have no upper bound on the size of the array. The size is implicitely define to be the maximum number of Events you have seen in the past before you to reach the flush limit.

I presume having an items array makes it easier to handle the response, as the bulk api response contains the per item information in the same order (and same number of entries) as this array has. By only storing items that have a callback one might be forced to correlate item in the array with items in the response.